src/structure_managers/adaptor_neighbour_list.hh

```diff
@@ -1006,49 +1021,51 @@ namespace rascal {
     // contiguously at the beginning of the list.
-    int ntot_atoms{0};
-    for (auto atom : *this->manager) {
-      int atom_tag = atom.get_atom_tag();
-      size_t cluster_index = this->manager->get_atom_index(atom_tag);
-      auto atom_type = atom.get_atom_type();
-      this->atom_tag_list.push_back(atom_tag);
+    // for (auto center : this->manager) {
+    //   auto atom_tag{center.get_atom_tag()};
+    //   size_t cluster_index = this->manager->get_atom_index(atom_tag);
+    //   auto atom_type = center.get_atom_type();
+    //   this->atom_tag_list.push_back(atom_tag);
+    //   this->atom_types.push_back(atom_type);
+    //   this->atom_index_from_atom_tag_list.push_back(cluster_index);
+    // }
+
+    // // fill up the tags for the missing center atoms. should never be
+    // accessed for (size_t i{this->n_centers}; i < this->n_atoms; ++i) {
+    //   this->atom_types.push_back(std::numeric_limits<int>::max());
+    //   this->atom_index_from_atom_tag_list.push_back(
+    //       std::numeric_limits<size_t>::max());
+    // }
+    for (size_t atom_tag{0}; atom_tag < this->n_atoms; ++atom_tag) {
+      auto atom_type = this->manager->get_atom_type(atom_tag);
+      auto cluster_index = this->manager->get_atom_index(atom_tag);
+      if (atom_tag < this->n_centers) {
+        this->atom_tag_list.push_back(atom_tag);
+      }
       this->atom_types.push_back(atom_type);
-      ntot_atoms++;
       this->atom_index_from_atom_tag_list.push_back(cluster_index);
     }
+
     // generate ghost atom tags and positions
-    for (auto atom : this->get_manager().with_ghosts()) {
-      auto pos = atom.get_position();
-      auto atom_type = atom.get_atom_type();
+    for (size_t atom_tag{0}; atom_tag < this->n_atoms; ++atom_tag) {
+      auto pos = this->manager->get_position(atom_tag);
+      auto atom_type = this->manager->get_atom_type(atom_tag);
 
       for (auto && p_image :
            internal::PeriodicImages<dim>{periodic_min, repetitions, ntot}) {
-        int ncheck{0};
-        for (auto i{0}; i < dim; ++i) {
-          ncheck += std::abs(p_image[i]);
-        }
-        // exclude cell itself
-        if (ncheck > 0) {
-          Vector_t pos_ghost{pos};
-
-          for (auto i{0}; i < dim; ++i) {
-            pos_ghost += cell.col(i) * p_image[i];
-          }
-
+        // exclude the original unit cell
+        if (not(p_image.array() == 0).all()) {
+          Vector_t pos_ghost{pos + cell * p_image.template cast<double>()};
           auto flag_inside =
               internal::position_in_bounds(ghost_min, ghost_max, pos_ghost);
 
           if (flag_inside) {
             // next atom tag is size, since start is at index = 0
-            auto new_atom_tag = this->n_centers + this->n_ghosts;
-            // get_size_with_ghosts();
+            auto new_atom_tag{this->n_atoms + this->n_ghosts};
             this->add_ghost_atom(new_atom_tag, pos_ghost, atom_type);
-            ntot_atoms++;
```

Comment:
Why is not needed anymore? I am a little confused.

Reply:
the variable is not needed anymore since n_ghosts does the same job